### ISSUE TITLE
[Permissions] Restarting App From Home Screen Icon

### ISSUE THREAD

**Issue by hvaughan3** [jamesmontemagno/Xamarin.Plugins#184]
I posted on Xamarin Forum but realized I should probably post it here instead.

It looks like something strange is happening and I am wondering if the permissions plugin has anything to do with it. It seems that before installing the plugin, pressing the home button to leave the application and then going and click the application button (NOT the shortcut) would bring the application back normally. But now it seems like when the application is already running and we go and click the application icon on the home screen (again, not the same as clicking the shortcut) the application seems to get restarted or at least does not go through the same process as if you were to open the multitask window view and click the running app's window.

So for example:

Open the application for first time ->
Hit home button ->
Click either the app shortcut or open the app from the multitask/window view ->
Everything works as expected

But if you instead:

Open the application for first time ->
Hit home button ->
Open the application from application icon

All of the logic in `OnAppearing()` that tells wether to force the user to login to the app again is skipped and it acts as if I have manually closed the application completely and restarted it.

Edit: After manually commenting out all of the code in `MainApplication.cs` and the `OnRequestPermissionsResult()` callback, it seems to be working correctly again and not restarting the app.

**Comment by jamesmontemagno** [comment]
I am not experiencing this at all on my nexus 6 P device.  I assume you are talking about Android.

My steps using Xamarin.Forms 2.0.x:

1.) Launch app, navigate to a page
2.) Hit home button
3.) Hit application drawer with all my apps
4.) Hit app again

Result is app is launched into the page it left off on.

The OnAppearing or OnDissapearing never get called on my page, which is correct, however the OnResume and OnSleep in my App.xaml.cs do.

What are your flags on your MainActivity?

Mine are: [Activity(Label = "AppName", Icon = "@drawable/ic_launcher", MainLauncher = true, ConfigurationChanges = ConfigChanges.ScreenSize | ConfigChanges.Orientation)]

**Comment by hvaughan3** [comment]
Thanks for the quick reply. I will try to investigate a little further when I have some time, maybe this weekend. My MainActivity flags are basically the same:

`[Activity(Label = "Mobile Beta", Icon = "@drawable/icon", Theme = "@style/LightTheme", MainLauncher = true, ConfigurationChanges = ConfigChanges.ScreenSize | ConfigChanges.Orientation)]`

I will look into it and see exactly what is being skipped. I am also seeing this on an old Samsung I8200 Galaxy S III mini VE, Android 4.2.2, which I realize does not use the new permissions model that 6.0 does.

Basically in `App.OnSleep()` I am setting a static variable to `DateTime.Now` then I am calculating the time in `App.OnStart()`. Having not looked into it fully, when I click the application icon, the app acts like it completely restarts or at least wipes out my static variables.

**Comment by jamesmontemagno** [comment]
OnStart() on forms will only get called when it starts for the first time.
OnResume gets called when brought back.

If you need to calculate time a static variable would not be the way to go
as Android itself can clear out you app or certain state. Use my settings
plugin to store the UTC time.

The permission plugin doesn't do anything special form pre lollipop at all.
You can add some breakpoints in the MainApplication class such as
OnTerminate and that is when Android terminates your app.

Best
James

**Comment by hvaughan3** [comment]
Sorry for the delay. I do plan to get back to you on this. Just been very busy. Keeping it closed for now is fine with me. Thank you.

**Comment by hvaughan3** [comment]
Finally figured out, more or less, what the issue is and it has nothing to do with your plugin (sorry about that)! I appreciate your help though.

Apparently I am running into the very old and still present (2009!) bug mostly in Android when the manufacturer customizes things. The bug reports are here:
[https://code.google.com/p/android/issues/detail?id=2373](https://code.google.com/p/android/issues/detail?id=2373)

and

[https://code.google.com/p/android/issues/detail?id=26658](https://code.google.com/p/android/issues/detail?id=26658)

The reason why commenting out stuff in MainApplication.cs seemed like it fixed the issue was probably because I you have to follow fairly specific steps to recreate this bug and I did not realize that at the time.

I implemented the following code to attempt to workaround the issue but this ends up crashing the app so not sure what to do next. This would go into `MainActivity.OnCreate()` right after `base.OnCreate(bundle);`:

    if(!IsTaskRoot) {
        Intent intent = Intent;
        string action = intent.Action;
        if(intent.HasCategory(Intent.CategoryLauncher) && action != null && action.Equals(Intent.ActionMain, System.StringComparison.OrdinalIgnoreCase)) {
            System.Console.WriteLine("\nIn APP.Droid.MainActivity.OnCreate() - Finishing Activity and returning since a second MainActivity has been created.\n");
            Finish();
            return;
        }
    }

**Comment by hvaughan3** [comment]
Just to chime back in here for someone else that might hit the issue I described above. It looks like Xamarin Android does not like it when you call `Finish()` before calling: `Xamarin.Forms.Forms.Init(this, bundle);` and `LoadApplication(new App());`.

So to finally get the above solution working correctly your `MainActivity.OnCreate()` would look like this:

    protected override void OnCreate(Bundle bundle) {
        base.OnCreate(bundle);

        Xamarin.Forms.Forms.Init(this, bundle);
        LoadApplication(new App());

        if(!IsTaskRoot) {
            Intent intent = Intent;
            string action = intent.Action;
            if(intent.HasCategory(Intent.CategoryLauncher) && action != null && action.Equals(Intent.ActionMain, System.StringComparison.OrdinalIgnoreCase)) {
                System.Console.WriteLine("\nIn APP.Droid.MainActivity.OnCreate() - Finishing Activity and returning since a second MainActivity has been created.\n");
                Finish();
                return; //Obviously this is not really required if you do not have anymore code below
            }
        }
    }

**Comment by jamesmontemagno** [comment]
This seems like this is some generic Android issue. I haven't really seen or experienced this.

What kind of Task are you specifying for your Main activity?

, LaunchMode = LaunchMode.SingleTask ?

Might be worth looking through: http://developer.android.com/guide/topics/manifest/activity-element.html#lmode
